// @remotion/media-parser/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.MediaParserInternals = exports.parseMedia = exports.IsAPdfError = exports.IsAnUnsupportedFileTypeError = exports.IsAnUnsupportedAudioTypeError = exports.IsAnImageError = exports.IsAGifError = void 0;
const aac_codecprivate_1 = require("./aac-codecprivate");
const event_emitter_1 = require("./create/event-emitter");
const create_iso_base_media_1 = require("./create/iso-base-media/create-iso-base-media");
const create_matroska_media_1 = require("./create/matroska/create-matroska-media");
const progress_tracker_1 = require("./create/progress-tracker");
const create_wav_1 = require("./create/wav/create-wav");
const with_resolvers_1 = require("./create/with-resolvers");
const log_1 = require("./log");
var errors_1 = require("./errors");
Object.defineProperty(exports, "IsAGifError", { enumerable: true, get: function () { return errors_1.IsAGifError; } });
Object.defineProperty(exports, "IsAnImageError", { enumerable: true, get: function () { return errors_1.IsAnImageError; } });
Object.defineProperty(exports, "IsAnUnsupportedAudioTypeError", { enumerable: true, get: function () { return errors_1.IsAnUnsupportedAudioTypeError; } });
Object.defineProperty(exports, "IsAnUnsupportedFileTypeError", { enumerable: true, get: function () { return errors_1.IsAnUnsupportedFileTypeError; } });
Object.defineProperty(exports, "IsAPdfError", { enumerable: true, get: function () { return errors_1.IsAPdfError; } });
var parse_media_1 = require("./parse-media");
Object.defineProperty(exports, "parseMedia", { enumerable: true, get: function () { return parse_media_1.parseMedia; } });
exports.MediaParserInternals = {
    createMatroskaMedia: create_matroska_media_1.createMatroskaMedia,
    createIsoBaseMedia: create_iso_base_media_1.createIsoBaseMedia,
    createWav: create_wav_1.createWav,
    Log: log_1.Log,
    IoEventEmitter: event_emitter_1.IoEventEmitter,
    makeProgressTracker: progress_tracker_1.makeProgressTracker,
    withResolvers: with_resolvers_1.withResolvers,
    withResolversAndWaitForReturn: with_resolvers_1.withResolversAndWaitForReturn,
    createAacCodecPrivate: aac_codecprivate_1.createAacCodecPrivate,
};
var version_1 = require("./version");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return version_1.VERSION; } });
