// @remotion/media-parser/dist/esm/web-fs.mjs
// src/writers/web-fs.ts
var createContent = async ({ filename }) => {
  const directoryHandle = await navigator.storage.getDirectory();
  directoryHandle.removeEntry(filename);
  const fileHandle = await directoryHandle.getFileHandle(filename, {
    create: true
  });
  const writable = await fileHandle.createWritable();
  let written = 0;
  let writPromise = Promise.resolve();
  const write = async (arr) => {
    await writable.write(arr);
    written += arr.byteLength;
  };
  const updateDataAt = async (position, data) => {
    await writable.seek(position);
    await writable.write(data);
    await writable.seek(written);
  };
  const remove = async () => {
    await directoryHandle.removeEntry(filename, {
      recursive: true
    });
  };
  const writer = {
    write: (arr) => {
      writPromise = writPromise.then(() => write(arr));
      return writPromise;
    },
    save: async () => {
      try {
        await writable.close();
      } catch {
      }
      const newHandle = await directoryHandle.getFileHandle(filename, {
        create: true
      });
      const newFile = await newHandle.getFile();
      return newFile;
    },
    getWrittenByteCount: () => written,
    updateDataAt: (position, data) => {
      writPromise = writPromise.then(() => updateDataAt(position, data));
      return writPromise;
    },
    waitForFinish: async () => {
      await writPromise;
    },
    remove
  };
  return writer;
};
var webFsWriter = {
  createContent
};
var canUseWebFsWriter = async () => {
  if (!("storage" in navigator)) {
    return false;
  }
  if (!("getDirectory" in navigator.storage)) {
    return false;
  }
  const directoryHandle = await navigator.storage.getDirectory();
  const fileHandle = await directoryHandle.getFileHandle("remotion-probe-web-fs-support", {
    create: true
  });
  const canUse = fileHandle.createWritable !== undefined;
  return canUse;
};
export {
  webFsWriter,
  canUseWebFsWriter
};
