// @remotion/media-parser/dist/boxes/webm/color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeMatroskaColorBytes = exports.parseColorSegment = void 0;
const truthy_1 = require("../../truthy");
const make_header_1 = require("./make-header");
const traversal_1 = require("./traversal");
const parseColorSegment = (colourSegment) => {
    const transferCharacteristics = (0, traversal_1.getTransferCharacteristicsSegment)(colourSegment);
    const matrixCoefficients = (0, traversal_1.getMatrixCoefficientsSegment)(colourSegment);
    const primaries = (0, traversal_1.getPrimariesSegment)(colourSegment);
    const range = (0, traversal_1.getRangeSegment)(colourSegment);
    return {
        transferCharacteristics: transferCharacteristics
            ? transferCharacteristics.value.value === 1
                ? 'bt709'
                : transferCharacteristics.value.value === 6
                    ? 'smpte170m'
                    : transferCharacteristics.value.value === 13
                        ? 'iec61966-2-1'
                        : null
            : null,
        matrixCoefficients: matrixCoefficients
            ? matrixCoefficients.value.value === 1
                ? 'bt709'
                : matrixCoefficients.value.value === 6
                    ? 'smpte170m'
                    : matrixCoefficients.value.value === 5
                        ? 'bt470bg'
                        : null
            : null,
        primaries: primaries
            ? primaries.value.value === 1
                ? 'bt709'
                : primaries.value.value === 6
                    ? 'smpte170m'
                    : primaries.value.value === 5
                        ? 'bt470bg'
                        : null
            : null,
        fullRange: (transferCharacteristics === null || transferCharacteristics === void 0 ? void 0 : transferCharacteristics.value.value) && (matrixCoefficients === null || matrixCoefficients === void 0 ? void 0 : matrixCoefficients.value.value)
            ? null
            : range
                ? Boolean(range === null || range === void 0 ? void 0 : range.value.value)
                : null,
    };
};
exports.parseColorSegment = parseColorSegment;
const makeMatroskaColorBytes = ({ transferCharacteristics, matrixCoefficients, primaries, fullRange, }) => {
    const rangeValue = transferCharacteristics && matrixCoefficients
        ? 3
        : fullRange === true
            ? 2
            : fullRange === false
                ? 1
                : 0;
    // https://datatracker.ietf.org/doc/draft-ietf-cellar-matroska/
    // 5.1.4.1.28.27
    const primariesValue = primaries === 'bt709'
        ? 1
        : primaries === 'smpte170m'
            ? 6
            : primaries === 'bt470bg'
                ? 5
                : 2;
    const transferChracteristicsValue = transferCharacteristics === 'bt709'
        ? 1
        : transferCharacteristics === 'smpte170m'
            ? 6
            : transferCharacteristics === 'iec61966-2-1'
                ? 13
                : 2;
    if (matrixCoefficients === 'rgb') {
        throw new Error('Cannot encode Matroska in RGB');
    }
    const matrixCoefficientsValue = matrixCoefficients === 'bt709'
        ? 1
        : matrixCoefficients === 'bt470bg'
            ? 5
            : matrixCoefficients === 'smpte170m'
                ? 6
                : 2;
    return (0, make_header_1.makeMatroskaBytes)({
        type: 'Colour',
        minVintWidth: null,
        value: [
            transferChracteristicsValue === 2
                ? null
                : {
                    type: 'TransferCharacteristics',
                    value: {
                        value: transferChracteristicsValue,
                        byteLength: null,
                    },
                    minVintWidth: null,
                },
            matrixCoefficientsValue === 2
                ? null
                : {
                    type: 'MatrixCoefficients',
                    value: {
                        value: matrixCoefficientsValue,
                        byteLength: null,
                    },
                    minVintWidth: null,
                },
            primariesValue === 2
                ? null
                : {
                    type: 'Primaries',
                    value: {
                        value: primariesValue,
                        byteLength: null,
                    },
                    minVintWidth: null,
                },
            {
                type: 'Range',
                value: {
                    value: rangeValue,
                    byteLength: null,
                },
                minVintWidth: null,
            },
        ].filter(truthy_1.truthy),
    });
};
exports.makeMatroskaColorBytes = makeMatroskaColorBytes;
